fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/DeleteResourceServiceImpl.java

```diff
@@ -0,0 +1,122 @@
+import org.fcrepo.kernel.api.services.DeleteResourceService;
+import org.fcrepo.persistence.api.PersistentStorageSession;
+import org.fcrepo.persistence.api.PersistentStorageSessionManager;
+import org.fcrepo.persistence.api.exceptions.PersistentStorageException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class mediates delete operations between the kernel and persistent storage layers
+ *
+ * @author dbernstein
+ */
+public class DeleteResourceServiceImpl implements DeleteResourceService {
+
+    private Logger LOGGER = LoggerFactory.getLogger(DeleteResourceService.class);
+
+    @Inject
+    private ContainmentIndex containmentIndex;
+
+    @Inject
+    private ResourceFactory resourceFactory;
+
+    @Inject
+    private DeleteResourceOperationFactory deleteResourceFactory;
+
+    @Inject
+    private PersistentStorageSessionManager psManager;
+
+    @Override
+    public void perform(final Transaction tx, final FedoraResource fedoraResource) {
+
+        final String fedoraResourceId = fedoraResource.getId();
+
+        try {
+            LOGGER.debug(format("deleting of %s", fedoraResourceId));
+            final PersistentStorageSession pSession = this.psManager.getSession(tx.getId());
+            deleteRecursivelyFromBottom(tx, pSession, fedoraResource);
+        } catch (final PersistentStorageException ex) {
+            throw new RepositoryRuntimeException(format("failed to delete resource %s", fedoraResourceId), ex);
+        }
+    }
+
+
+    private void deleteRecursivelyFromBottom(final Transaction tx, final PersistentStorageSession pSession,
+                                             final FedoraResource fedoraResource) throws PersistentStorageException {
+
+        final String fedoraId = fedoraResource.getId();
+
+        if (fedoraResource instanceof Container) {
+            final Stream<String> children = containmentIndex.getContainedBy(tx, fedoraResource);
+            children.forEach(childResourceId -> {
+                try {
+                    final FedoraResource res = resourceFactory.getResource(tx, childResourceId);
+                    deleteRecursivelyFromBottom(tx, pSession, res);
+                } catch (final PathNotFoundException ex) {
+                    LOGGER.warn(format("Path not found for %s: ignoring.", childResourceId));
+                } catch (final PersistentStorageException ex) {
+                    throw new RepositoryRuntimeException(format("failed to delete resource %s", fedoraId), ex);
+                }
+            });
+        }
+
+        //delete described resource if binary
+        if (fedoraResource instanceof Binary) {
+            delete(pSession, fedoraResource.getDescribedResource().getId());
+        }
+
+        //delete the acl if this is not the acl
+        if (!fedoraResource.isAcl()) {
+            final FedoraResource acl = fedoraResource.getAcl();
+            if (fedoraResource.getAcl() != null) {
```

Comment:
use the `acl` variable that was just declared instead of retrieving again. Same with the next line.

Reply:
Re :
>  If you delete a AG, does the whole OCFL object go away?

Good question.  We will continue to support Tombstones no?  In that case, the OCFL Object would only go away once all the tombstone's were deleted.   

Also question:   

if `/my-ag`  is deleted, would we expect  `/my-ag/binary1/fcr:tombstone` to exist?
Then if we delete `/my-ag/fcr:tombstone` would we expect all tombstones of the children also to be deleted,  ie `/my-ag/binary1/fcr:tombstone`?

Reply:
I'm going to add this question to https://wiki.duraspace.org/display/FF/Decisions+and+Open+Questions

Reply:
@dbernstein The current behavior in fcrepo5 is that a GET to the child resource of a deleted container tells you where the tombstone of the deleted parent is, rather than having tombstones for all the children.